sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala

```diff
@@ -642,81 +643,77 @@ case class GpuSum(child: Expression,
-    case _ => aggBufferAttributes
+    case dt: DecimalType =>
+      Seq(GpuCheckOverflow(postUpdateSum, dt, !failOnErrorOverride), postUpdateIsEmpty)
+    case _ => postUpdateAttr
+  }
+
+  // output of GpuSum
+  private lazy val sum = AttributeReference("sum", resultType)()
+  // Used for Decimal overflow detection
+  private lazy val isEmpty = AttributeReference("isEmpty", BooleanType)()
+  override lazy val aggBufferAttributes: Seq[AttributeReference] = resultType match {
+    case _: DecimalType => sum :: isEmpty :: Nil
+    case _ => sum :: Nil
   }
 
   override lazy val preMerge: Seq[Expression] = resultType match {
-    case _: DecimalType => Seq(cudfSum, isEmpty, GpuIsNull(cudfSum))
+    case _: DecimalType => Seq(sum, isEmpty, GpuIsNull(sum))
     case _ => aggBufferAttributes
   }
 
-  override def mergeBufferAttributes: Seq[AttributeReference] = postMergeAttr
-
   // To be able to do decimal overflow detection, we need a CudfSum that does **not** ignore nulls.
   // Cudf does not have such an aggregation, so for merge we have to work around that similar to
   // what happens with isEmpty
-  override lazy val mergeExpressions: Seq[Expression] = resultType match {
-    case _: DecimalType => Seq(new CudfSum(cudfSum), new CudfMin(isEmpty), new CudfMax(isOverflow))
-    case _ => Seq(new CudfSum(cudfSum))
+  override lazy val mergeAggregates: Seq[CudfAggregate] = resultType match {
+    case _: DecimalType =>
+      Seq(new CudfSum(resultType), new CudfMin(BooleanType), new CudfMax(BooleanType))
+    case _ => Seq(new CudfSum(resultType))
   }
 
-  override lazy val postMergeAttr: Seq[AttributeReference] = resultType match {
-    case _: DecimalType => cudfSum :: isEmpty :: isOverflow :: Nil
-    case _ => aggBufferAttributes
-  }
+  private lazy val postMergeSum = mergeAggregates.head.attr
+  private lazy val postMergeIsEmpty = mergeAggregates(1).attr
+  private lazy val postMergeIsOverflow = mergeAggregates(2).attr
 
   override lazy val postMerge: Seq[Expression] = resultType match {
     case _: DecimalType =>
-      Seq(GpuIf(isOverflow, GpuLiteral.create(null, resultType), cudfSum), isEmpty)
-    case _ => aggBufferAttributes
+      Seq(GpuIf(postMergeIsOverflow, GpuLiteral.create(null, resultType), postMergeSum),
+        postMergeIsEmpty)
+    case _ => postMergeAttr
   }
 
   override lazy val evaluateExpression: Expression = resultType match {
     case dt: DecimalType =>
-      GpuCheckOverflowAfterSum(cudfSum, isEmpty, dt, !failOnErrorOverride)
-    case _ => cudfSum
-  }
-
-  override lazy val aggBufferAttributes: Seq[AttributeReference] = resultType match {
-    case _: DecimalType => cudfSum :: isEmpty :: Nil
-    case _ => cudfSum :: Nil
-  }
-
-  override lazy val initialValues: Seq[GpuLiteral] = resultType match {
-    case _: DecimalType => Seq(zeroDec, GpuLiteral(true, BooleanType))
-    case _ => Seq(GpuLiteral(null, resultType))
+      GpuCheckOverflowAfterSum(sum, isEmpty, dt, !failOnErrorOverride)
```

Comment:
a side note, `GpuCheckOverflowAfterSum` has the logic reversed on when it should keep nulls or when it should throw in its impl.

Reply:
Sorry I should have had a better check. I didn't want to spend an extra min of test time just to check overflow for decimals on SUM, but I probably should have done it.